torch/_inductor/kernel/mm.py

```diff
@@ -492,6 +493,9 @@ def tuned_mixed_mm(mat1, mat2, mat2_dtype):
         CUTLASSGemmTemplate.add_cutlass_gemm_choices(
             choices, layout, [mat1, mat2], fuseable=True, non_fuseable=True
         )
+        CUTLASS2xGemmTemplate.add_cutlass_gemm_choices(
```

Comment:
Can both of these be used unconditionally? Or does there need to be a branch on the cutlass version? 

If we can use both, do we need both?

Reply:
Edit, from @alexsamardzic comment I see this is adding back support for Ampere, so if the existing template does not work for ampere should there be a branch on SM verson to decide which to add?

Reply:
Roughly speaking: `add_cutlass_gemm_choices()` will go through all configurations offered by CUTLASS, and will then `filter_op()` on each one.  For `CUTLASSGemmTemplate`, `filter_op()` will eliminate operations that are not for Hopper, and for `CUTLASS2xGemmTemplate`, it will eliminate operations that are not for Ampere.  So the only harm here, as is, is going through all the configurations twice - but that's just another reason to unify these two classes.

Reply:
I would think that it makes sense to factor out the gen_ops() and filter_op() methods from CutlassGEMMTemplate and CUTLASS2xGemmTemplate to a shared module and make them functions, not methods. Then you can pass in a flag that controls that. @alexsamardzic is that the kind of planned refactoring you had in mind already?

Reply:
(This comment applies to your suggestions below too.)  The main point is not in commonalities, but exactly that differences between the two classes are actually minor.  Thus, my questions is: are you still at the position that we need to have two separate classes, plus a base class?  Frankly, I think single class with some if-else statements, like what was the case before the major refactoring, would be simpler.  Plus, it would avoid going through all the ops offered by CUTLASS generator twice, in cases when we want to support both CUTLASS 2.x and CUTLASS 3.x architecture kernels for auto-tuning the same operator.

Reply:
I can understand where you're coming from. It might appear simpler and more compact to write at the moment, but it definitely impacts readability and testability ( and thereby maintainability ) very badly and that's not to be underestimated. So, yes, I still think that it's better to have two classes. But that's a bit of Software Design philosophy. I personally prioritize readability, modularity, testability and orthogonality in software design over the DRY principle, but I know that opinions on that differ.

Going through all ops twice can be solved differently ( sort them into separate lists / sets just once and remember these ). It's inefficient anyway at the moment. Cutlass 3.x will certainly evolve a lot in the future, while Cutlass 2.x likely won't, so I would not just judge what's easy and what is not by how the code looks at the moment. 

You could factor out shared code into a shared parent class or an utility module if that helps.

Reply:
OK.  I'm on leave now for several weeks, but as soon as I'm back, I'll refactor the code and will ping you for a review then.

Reply:
Same for me, I will be back in a bit more than 2 weeks. Sorry I had to delay this ..

Reply:
@alexsamardzic: Yes, I still think it's preferable to have two classes with shared code factored out into a common base class or into shared utility functions instead of a single class with if-else statements. That's the approach which enables independent / orthogonal future developments and keeps the implementation simple, readable and testable going forward.